hack/e2e.go

```diff
@@ -126,109 +129,116 @@ func main() {
 			}
 		}
 		if *up {
-			if !Up() {
-				log.Fatal("Error starting e2e cluster. Aborting.")
-			}
+			doOrDie("starting e2e cluster", Up())
 		} else if *push {
-			if !finishRunning("push", exec.Command("./hack/e2e-internal/e2e-push.sh")) {
-				log.Fatal("Error pushing e2e cluster. Aborting.")
-			}
+			doOrDie("pushing e2e clsuter", finishRunning("push", exec.Command("./hack/e2e-internal/e2e-push.sh")))
 		}
 	}
 
 	upResources := ""
 	if *checkLeakedResources {
-		upResources = ListResources()
+		upResources, err = ListResources()
+		errs = appendError(errs, err)
 	}
 
-	success := true
-
 	if *ctlCmd != "" {
 		ctlArgs := strings.Fields(*ctlCmd)
 		os.Setenv("KUBE_CONFIG_FILE", "config-test.sh")
-		ctlSuccess := finishRunning("'kubectl "+*ctlCmd+"'", exec.Command("./cluster/kubectl.sh", ctlArgs...))
-		success = success && ctlSuccess
+		errs = appendError(errs, finishRunning("'kubectl "+*ctlCmd+"'", exec.Command("./cluster/kubectl.sh", ctlArgs...)))
 	}
 
 	if *upgradeArgs != "" {
-		upgradeSuccess := UpgradeTest(*upgradeArgs)
-		success = success && upgradeSuccess
+		errs = appendError(errs, UpgradeTest(*upgradeArgs))
 	}
 
 	if *test {
 		if *skewTests {
-			skewSuccess := SkewTest()
-			success = success && skewSuccess
+			errs = appendError(errs, SkewTest())
 		} else {
-			testSuccess := Test()
-			success = success && testSuccess
+			errs = appendError(errs, Test())
 		}
 	}
 
 	if *kubemark {
-		kubeSuccess := KubemarkTest()
-		success = success && kubeSuccess
+		errs = appendError(errs, KubemarkTest())
 	}
 
 	if *down {
-		if !success && *dump != "" {
+		if errs != nil && *dump != "" {
 			DumpClusterLogs(*dump)
 		}
-		tearSuccess := TearDown()
-		success = success && tearSuccess
+		errs = appendError(errs, TearDown())
 	}
 
 	if *checkLeakedResources {
 		log.Print("Sleeping for 30 seconds...") // Wait for eventually consistent listing
 		time.Sleep(30 * time.Second)
-		DiffResources(beforeResources, upResources, ListResources(), *dump)
+		afterResources, err := ListResources()
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			errs = appendError(errs, DiffResources(beforeResources, upResources, afterResources, *dump))
+		}
 	}
 
-	if !success {
-		os.Exit(1)
+	if errs != nil {
```

Comment:
Since this is a slice `if len(errs) != 0` might be better. It works with both nil and empty slices.